### ISSUE TITLE
Skipping all strings in XML

### ISSUE THREAD

**Issue by zallaevan** [mythkiven/mkTranslate#6]
Although I've tried with Google and Youdao, it seems that it isn't capable of translating XML archives. The following command is given:
`translate -p ./fr.xml -d 'zh-hans' -c ''youdao -s 'fr'`
It is supposed to take the strings in french and convert them into simplified Chinese, but it seems that it only skips them. I attach some screenshots below.
![image](https://user-images.githubusercontent.com/31564969/87157731-f6e50380-c2be-11ea-8db1-8b939a27e33e.png)
![image](https://user-images.githubusercontent.com/31564969/87157753-019f9880-c2bf-11ea-92c7-434028ad4602.png)

**Comment by XilinJia** [comment]
skipping English too.  Seems totally useless.

**Comment by zallaevan** [comment]
> skipping English too. Seems totally useless.

Yeah, no clue why this happens. Maybe because it is not the "default" format? Nevertheless, I've put it through an XML checker and it seems that it's all ok.

**Comment by XilinJia** [comment]
@zallaevan So it works for you now?

Meanwhile I found this one (https://github.com/Ra-Na/GTranslate-strings-xml) works fine for me after a simple edit of my xml file, about which I filed an issue there.

**Comment by zallaevan** [comment]
@LeoK987 No, it doesn't work for me, at least. I said that the online XML checker says that the syntax is correct, but the program doesn't function properly.
Thank you for suggesting that, I'll give it a look. Cheers.

**Comment by mythkiven** [comment]
Hi Evan, thanks for the detailed report and screenshots!

I see **two likely causes** for all strings being skipped:

### 1. Command typo

Your command shows:
```bash
translate -p ./fr.xml -d 'zh-hans' -c ''youdao -s 'fr'
```

Note `-c ''youdao` — there is an **extra single quote** before `youdao`. It should be:

```bash
translate -p ./fr.xml -d 'zh-hans' -c youdao -s fr
```

### 2. How XML matching works

mkTranslate does **not** parse XML as a tree. It uses regex per line:

```python
r">\s*(.+?)\s*</string>"
```

A line is **skipped** (logs `skip: ...`) when:
- The string value is empty
- Content spans multiple lines
- The line contains `<![CDATA[...]]>` or inner HTML tags that break the pattern
- The `>` … `</string>` pattern doesn't match

For French files, also verify:
- File encoding is **UTF-8**
- Each translatable string is on a **single line** like:
  ```xml
  <string name="foo">Bonjour</string>
  ```

### Debugging steps

Run again and watch console output — you'll see either:
- `original:...` → `translated:...` (success)
- `skip: ...` (regex didn't match)

Try Google channel too:
```bash
translate -p ./fr.xml -d zh-hans -c google -s fr
```

### If strings use CDATA or HTML

Example that **won't work** with current regex:
```xml
<string name="x"><![CDATA[Hello <b>world</b>]]></string>
```

For those files, you'd need to strip/format first, or use a proper XML parser (feature improvement).

Please retry with the fixed `-c youdao` command and share a sample line that gets skipped — I can tell you exactly why.

**Comment by mythkiven** [comment]
Resolved in mkTranslate v2.0 — see comment below.

**Comment by mythkiven** [comment]
Fixed in **v2.0**.

Root causes addressed:
1. **Command typo** in original report (`-c ''youdao`) — now documented.
2. **Regex parser replaced** with `ElementTree` — CDATA, multi-line, and nested text now work.

Try:
```bash
translate -p ./fr.xml -d zh-hans -c youdao -s fr
```

Closing as resolved.
